pyramid/tierFS.go

```diff
@@ -0,0 +1,364 @@
+func (tfs *TierFS) Open(namespace, filename string) (*ROFile, error) {
+	if err := validateArgs(namespace, filename); err != nil {
+		return nil, err
+	}
+
+	fileRef := tfs.newLocalFileRef(namespace, filename)
+	fh, err := os.Open(fileRef.fullPath)
+	if err == nil {
+		return tfs.openFile(fileRef, fh)
+	}
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("open file: %w", err)
+	}
+
+	fh, err = tfs.readFromBlockStorage(fileRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return tfs.openFile(fileRef, fh)
+}
+
+// openFile converts an os.File to pyramid.File and updates the eviction control.
+func (tfs *TierFS) openFile(fileRef localFileRef, fh *os.File) (*ROFile, error) {
+	stat, err := fh.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("file stat: %w", err)
+	}
+
+	// No need to check for the store output here,
+	// we already have the file.
+	tfs.eviction.store(fileRef.fsRelativePath, stat.Size())
+	return &ROFile{
+		fh:       fh,
+		rPath:    fileRef.fsRelativePath,
+		eviction: tfs.eviction,
+	}, nil
+}
+
+// readFromBlockStorage reads the referenced file from the block storage
+// and places it in the local FS for further reading.
+// It returns a file handle to the local file.
+func (tfs *TierFS) readFromBlockStorage(fileRef localFileRef) (*os.File, error) {
+	reader, err := tfs.adaptor.Get(tfs.objPointer(fileRef.namespace, fileRef.filename), 0)
+	if err != nil {
+		return nil, fmt.Errorf("read from block storage: %w", err)
+	}
+	defer reader.Close()
+
+	writer, err := os.Create(fileRef.fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating file: %w", err)
+	}
+
+	if _, err := io.Copy(writer, reader); err != nil {
+		return nil, fmt.Errorf("copying date to file: %w", err)
+	}
+	if err := writer.Close(); err != nil {
+		return nil, fmt.Errorf("writer close: %w", err)
+	}
+
+	fh, err := os.Open(fileRef.fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %w", err)
+	}
+	return fh, nil
+}
+
+func validateArgs(namespace, filename string) error {
+	if err := validateNamespace(namespace); err != nil {
+		return err
+	}
+	return validateFilename(filename)
+}
+
+var (
+	errSeparatorInFS   = errors.New("path contains separator")
+	errPathInWorkspace = errors.New("file cannot be located in the workspace")
+	errEmptyDirInPath  = errors.New("file path cannot contain an empty directory")
+)
+
+func validateFilename(filename string) error {
+	if strings.HasPrefix(filename, workspaceDir+string(os.PathSeparator)) {
+		return errPathInWorkspace
+	}
+	if strings.Contains(filename, strings.Repeat(string(os.PathSeparator), 2)) {
+		return errEmptyDirInPath
+	}
+	return nil
+}
+
+func validateNamespace(ns string) error {
+	if strings.ContainsRune(ns, os.PathSeparator) {
+		return errSeparatorInFS
+	}
+	return nil
+}
+
+// relativePath is the path of the file under TierFS
+type relativePath string
+
+// localFileRef consists of all possible local file references
+type localFileRef struct {
```

Comment:
I suggest that the two last properties will be methods on localFileRef

Reply:
Good point. But then I'll have to keep a pointer to tfs, which is a bit messy.
I don't feel strongly about it, willing to change if you want.